drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+
+		if (page_count > max_page_desc) {
+			page_count_pass = max_page_desc;
+			page_count = page_count - max_page_desc;
+			cond_start = cond_start + page_count_pass;
+		} else {
+			page_count_pass = page_count;
+			page_count = page_count - page_count_pass;
+		}
+		for (int index = 0; index < page_count_pass; index++) {
+			cdns_nand_cdma_prepare(NF_TDEF_DEV_NUM, row_address, buffer,
+					       (ctype + index), (page_count_pass - index),
+					       DMA_MS_SEL, VOL_ID, (cdma_desc + index));
+		}
+		ret = cdns_nand_send(base_address, (char *)cdma_desc, NF_TDEF_TRD_NUM);
+
+		if (ret != 0) {
+			k_free(cdma_desc);
+			return ret;
+		}
+
+		if (ctype != CNF_CMD_ERASE) {
+			buffer = buffer + (max_page_desc * params->page_size);
+		}
+
+		ret = cdns_transfer_complete(cdma_desc, params);
+
+		if (ret != 0) {
+			k_free(cdma_desc);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+/**
+ * Perform a CDMA write operation for the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_page_number The starting page number for the write operation.
+ * @param buffer The buffer containing the data to be written.
+ * @param page_count The number of pages to be written.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_cdma_write(struct cadence_nand_params *params, uint32_t start_page_number,
+				char *buffer, uint32_t page_count)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+	int ret;
+
+	cdma_desc = cdma_allocate_desc_memory(page_count, CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT);
+
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
+	ret = cdns_cdma_desc_transfer_finish(params, cdma_desc, page_count,
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_WR,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
+
+	k_free(cdma_desc);
+	return 0;
+}
+
+/**
+ * Perform a CDMA read operation for the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_page_number The starting page number for the read operation.
+ * @param buffer The buffer to store the read data.
+ * @param page_count The number of pages to be read.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_cdma_read(struct cadence_nand_params *params, uint32_t start_page_number,
+			       char *buffer, uint32_t page_count)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+	int ret;
+
+	cdma_desc = cdma_allocate_desc_memory(page_count, CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT);
+
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
+	ret = cdns_cdma_desc_transfer_finish(params, cdma_desc, page_count,
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_RD,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
```

Comment:
There is double free everywhere where you do that after calling `cdns_cdma_desc_transfer_finish`, because on error conditions in `cdns_cdma_desc_transfer_finish` you release the buffer first time.